Transformer Implementation

yagera opened the pull request.

yagera committed:
refactor(attention): rename h/n_heads to num_heads

yagera committed:
fix(encoder): correct initialization parameters and remove self_attention arg

yagera committed:
fix(decoder): standardize layer initialization and naming (num_heads param)

yagera committed:
refactor(embeddings): reorder arguments

yagera committed:
Feat: Implement Transformer

yagera committed:
Feat: add torchinfo model summary to Transformer

yagera committed:
Feat(dataset): add dataset loading and processing logic

yagera committed:
Feat(config): add model configuration and paths setup

yagera committed:
Feat(training): implement main training loop with WandB integration

KORALLLL commented on models/modules/configuration.py:
Я советую создать отдельную папку, где будет хранится конфиг модели. Для работы с конфигами лучше использовать hydra https://hydra.cc/docs/intro/ 

Сам по себе фреймворк очень удобный, плюсом при каждом запуске он сохраняет информацию о текущем конфиге. В общем посмотри, изучи и начни его использовать вместо питоновского словаря. Но, если разобраться, то можно будет получить конфинг из .yaml файла и работать с ним, как с питоновким словарём

KORALLLL commented on models/modules/dataset.py:
```diff
@@ -0,0 +1,114 @@
+        tgt_text = item[self.tgt_lang]
+        src_clean = self.tokenizer.clean_text(src_text, self.src_lang)
+        tgt_clean = self.tokenizer.clean_text(tgt_text, self.tgt_lang)
+        src_token_to_id = getattr(self.tokenizer, f'{self.src_lang}_token_to_id')
+        src_vocab = getattr(self.tokenizer, f'{self.src_lang}_vocab')
+        tgt_token_to_id = getattr(self.tokenizer, f'{self.tgt_lang}_token_to_id')
+        tgt_vocab = getattr(self.tokenizer, f'{self.tgt_lang}_vocab')
```
может быть стоит эти методы перенести в конструктор? Я не вижу в них зависимости от текущего высказывания, в то время как постоянно выполнение их при итерирование является лишним

KORALLLL commented on models/modules/decoder.py:
```diff
@@ -3,30 +3,42 @@
 
+
 class Decoder(nn.Module):
-    def __init__(self, d_model: int, n_layers: int, n_heads: int, d_ff: int, dropout: float):
+    def __init__(self, d_model: int, layers: nn.ModuleList):
         super().__init__()
-        self.layers = nn.ModuleList([DecoderBlock(d_model, n_heads, d_ff, dropout) for _ in range(n_layers)])
+        self.layers = layers
```
верни инициализацию по гиперпараметрам, а не по переданным слоям

KORALLLL commented on models/modules/decoder.py:
```diff
@@ -3,30 +3,42 @@
     tgt_mask = torch.tril(torch.ones(seq_len, seq_len)).unsqueeze(0).unsqueeze(0)
-    decoder = Decoder(d_model=d_model, n_layers=n_layers, n_heads=n_heads, d_ff=d_ff, dropout=dropout)
+
+    decoder = Decoder(
+        d_model=d_model,
+        num_layers=num_layers,
+        num_heads=num_heads,
+        d_ff=d_ff,
+        dropout=dropout
+    )
+
```
Декодер кстати, не принимает такие парамтры, верни их))

KORALLLL commented on models/modules/encoder.py:
```diff
@@ -4,9 +4,9 @@
 class Encoder(nn.Module):
-    def __init__(self,d_model: int,n_layers: int,n_heads: int,d_ff: int,dropout: float):
+    def __init__(self, d_model: int, layers: nn.ModuleList):
         super().__init__()
-        self.layers = nn.ModuleList([EncoderBlock(d_model, n_heads, d_ff, dropout)for _ in range(n_layers)])
+        self.layers = layers
```
Верни инициализацию по гиперпараметрам

KORALLLL commented on models/modules/encoder.py:
```diff
@@ -16,11 +16,17 @@ def forward(self, x: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
     d_ff = 2048
     dropout = 0.1
-    encoder = Encoder(d_model=d_model,n_layers=n_layers,n_heads=n_heads,d_ff=d_ff,dropout=dropout)
+    encoder = Encoder(
+        d_model=d_model,
+        num_layers=num_layers,
+        num_heads=num_heads,
+        d_ff=d_ff,
+        dropout=dropout
+    )
```
то же самое, что и с декодером

KORALLLL commented on models/modules/multihead_attention.py:
```diff
@@ -4,54 +4,39 @@
-        key = self.w_k(k).view(batch_size, -1, self.h, self.d_k).transpose(1, 2)
-        value = self.w_v(v).view(batch_size, -1, self.h, self.d_k).transpose(1, 2)
-
-        x, attn = MultiHeadAttention.attention(query, key, value, mask, self.dropout)
-
+        q = self.w_q(q).view(batch_size, -1, self.num_heads, self.d_k).transpose(1, 2)
+        k = self.w_k(k).view(batch_size, -1, self.num_heads, self.d_k).transpose(1, 2)
+        v = self.w_v(v).view(batch_size, -1, self.num_heads, self.d_k).transpose(1, 2)
+        scores = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(self.d_k)
+        attn = torch.softmax(scores, dim=-1)
+        attn = self.dropout(attn)
+        x = torch.matmul(attn, v)
```
куда маскирование пропало?

KORALLLL commented on models/modules/transformer.py:
```diff
@@ -0,0 +1,165 @@
+from torchinfo import summary
+
+
+class Transformer(nn.Module):
+    def __init__(self, encoder: Encoder, decoder: Decoder,
+                 src_embed: InputEmbeddings, tgt_embed: InputEmbeddings,
+                 src_pos: PositionalEncoding, tgt_pos: PositionalEncoding,
+                 projection_layer: ProjectionLayer):
+        super().__init__()
+        self.encoder = encoder
+        self.decoder = decoder
+        self.src_embed = src_embed
+        self.tgt_embed = tgt_embed
+        self.src_pos = src_pos
+        self.tgt_pos = tgt_pos
+        self.projection_layer = projection_layer
```
не надо так делать, пожалуйста... Надо каждый из этих полей инициализировать внутри конструктора, не надо их передавать при инициализации

KORALLLL commented on models/modules/transformer.py:
```diff
@@ -0,0 +1,165 @@
+        return self.projection_layer(x)
+
+
+def build_transformer(
+        src_vocab_size: int,
+        tgt_vocab_size: int,
+        src_seq_len: int,
+        tgt_seq_len: int,
+        d_model: int = 512,
+        num_layers: int = 6,
+        num_heads: int = 8,
+        dropout: float = 0.1,
```
Вот эту всю замечательную логику инициализации надо перенести в конструктор

KORALLLL commented on models/modules/train.py:
```diff
@@ -0,0 +1,157 @@
+        d_ff=config["d_ff"]
+    ).to(device)
+
+    wandb.watch(model, log="all", log_freq=100)
```
о, это прикольная штука. Объясни, пожалуйста, зачем её используешь)

KORALLLL commented on models/modules/train.py:
```diff
@@ -0,0 +1,157 @@
+
+    for epoch in range(epoch, config["num_epochs"]):
+        model.train()
+        progress_bar = tqdm(train_ds, desc=f"Epoch {epoch + 1}/{config['num_epochs']}")
```
По-хорошему прогресс бар пихать в декларации цикла.

KORALLLL replied:
актуально

KORALLLL commented on models/modules/train.py:
```diff
@@ -0,0 +1,157 @@
+
+    epoch, global_step = load_checkpoint(config, model, optimizer)
+
+    for epoch in range(epoch, config["num_epochs"]):
```
вот тут бы тоже прогресс бар не помешал бы

KORALLLL replied:
актуально

KORALLLL commented on models/modules/train.py:
```diff
@@ -0,0 +1,157 @@
+            optimizer.step()
+            optimizer.zero_grad()
+
+            if global_step % config["log_interval"] == 0:
+                log_data = {
+                    "train/loss": loss.item(),
+                    "lr": optimizer.param_groups[0]['lr'],
+                    "epoch": epoch,
+                    "step": global_step
+                }
```
такое лучше логировать не каждые несколько степов, а каждый степ

KORALLLL replied:
и можно сразу в wandb

KORALLLL commented on models/modules/train.py:
```diff
@@ -0,0 +1,157 @@
+        for batch in tqdm(val_loader, desc="Validating"):
+            inputs = {k: v.to(device) for k, v in batch.items() if k != 'src_text' and k != 'tgt_text'}
+            encoder_output = model.encode(inputs['encoder_input'], inputs['encoder_mask'])
+            decoder_output = model.decode(encoder_output, inputs['encoder_mask'], inputs['decoder_input'], inputs['decoder_mask'])
+            proj_output = model.project(decoder_output)
```
это не совсем корректнаяя валидация. Во-первых, на валидации надо прям декодировать последовательность. Во-вторых, считать метрики, а не только лосс

yagera committed:
Refactor(encoder): revert hyperparameter initialization

yagera committed:
Refactor(decoder): revert hyperparameter initialization

yagera committed:
Feat(config): add hydra config for transformer training

yagera committed:
refactor: remove configuration.py

yagera committed:
refactor: remove utils from configuration.py to train.py

yagera committed:
feat: enhance validation with metrics and logging

- Remove wandb.watch
- Add progress bars for epochs
- Log metrics every step
- Implement BLEU and exact match validation metrics
- Add reference translations to example logging

yagera committed:
feat: add attention masking

yagera committed:
perf: move token vocab to dataset constructor

yagera committed:
refactor: move component initialization into class constructor

KORALLLL commented on models/configs/config.yaml:
очень хороший конфиг

yagera committed:
refactor: change using build_transformer to Transformer

yagera committed:
feat: add progress bar for epoches and batches

KORALLLL commented on models/modules/train.py:
```diff
@@ -0,0 +1,214 @@
+
+    epoch_progress = tqdm(range(epoch, cfg.training.num_epochs),desc="Training",position=0)
+
+    for epoch in epoch_progress:
```
прогресс бар

KORALLLL commented on models/modules/train.py:
```diff
@@ -0,0 +1,214 @@
+        epoch_progress.set_description(f"Epoch {epoch + 1}/{cfg.training.num_epochs}")
+        model.train()
+
+        batch_progress = tqdm(train_ds,desc=f"Batches",leave=False,position=1)
+
+        for batch in batch_progress:
```
прогресс бар по-другому
```python
for i in tqdm(train_ds, desc="..", leave=False):
```

yagera committed:
fix: update relative path

yagera committed:
feat: add new tokens

yagera committed:
fix: correct vocabulary paths

yagera committed:
feat: add translate method to Transformer class

yagera committed:
feat: add progress bars for training epochs and batches

yagera committed:
fix: replace decode with decode_ids in translation logging

yagera committed:
fix: pass id_to_token dictionary to decode_ids in translation logging

yagera committed:
fix: improve translate method

yagera committed:
fix: log translations per epoch in wandb

yagera committed:
fix: add version_base to hydra.main to suppress warning

yagera committed:
fix: correct loss function input size in validation and training

yagera committed:
fix: limit dataset to 50,000 sentences

yagera committed:
fix: handle batch processing in translate method

yagera committed:
add: console logging for translations in log_translations

yagera committed:
fix: use correct start and end token IDs in translate

yagera committed:
fix: use correct start and end token IDs in run_validation translate

yagera committed:
update: adjust lr, batch_size, and num_epochs in config

yagera committed:
fix: support new tokens in tokenizer

yagera committed:
fix: increase lr to 0.001

yagera committed:
remove: exact_match metric from run_validation

yagera committed:
feat: add one more example in log_translations

yagera committed:
fix: make examples more correct

yagera committed:
fix: capitalize first letter of translations in decode_ids

yagera committed:
fix: improve decode_ids for better punctuation handling

yagera committed:
fix: improve decode_ids to remove _ed and handle subtokens

yageraaa commented:
Почитал, что можно улучшить или добавить:

**1. Температура (Temperature Sampling)**

Это техника, которая смягчает или усиливает распределение вероятностей токенов при генерации текста. Она применяется к логитам (выходам модели перед softmax) в методе translate, чтобы сделать выбор токенов более или менее случайным:

- Температура < 1.0 (например, 0.7): Делает распределение более "острым", усиливая вероятности более вероятных токенов, но сохраняя некоторое разнообразие. Это уменьшает повторения (например, "papapapa") и делает переводы более осмысленными.

- Температура = 1.0: Оставляет распределение без изменений (стандартный softmax).

- Температура > 1.0 (например, 1.5): Делает распределение более "плоским", увеличивая вероятность менее частых токенов, что может привести к более креативным, но менее точным переводам.

_Преимущества:_

- Уменьшает повторяющиеся последовательности.

- Увеличивает разнообразие переводов, делая их более естественными.

**2.  Beam Search** 

Beam search отслеживает k наиболее вероятных последовательностей токенов (beam width) на каждом шаге, выбирая лучшую по суммарной вероятности в конце. Это более сложный метод, чем жадное декодирование или сэмплирование, но он часто даёт более точные переводы.

_Преимущества:_

- Значительно улучшает качество перевода, особенно для длинных предложений.

- Уменьшает вероятность "застревания" на локальных максимумах, как в жадном декодировании.

- Хорошо подходит для перевода, где важна точность.

**3. Warmup для скорости обучения**

Warmup постепенно увеличивает скорость обучения (lr) в первые несколько тысяч шагов, что стабилизирует обучение трансформеров, особенно с высоким lr (например, 0.001 или 0.002). Это помогает избежать резких изменений весов в начале обучения.

_Преимущества:_

- Улучшает стабильность обучения.
- Позволяет безопасно использовать более высокий lr.
- Ускоряет сходимость, что важно с учётом длинных последовательностей.

yageraaa commented:
Думаю, что температура никак не навредит, да и встраивается легко. А насчет остального завтра тогда.

yagera committed:
add: temperature sampling (0.7) to translate for better diversity

yageraaa commented:
Также встречается очень много повторений по типу: The of of of. Поэтому добавлю top-k выборку с k=10. И тогда у меня будет выбор 10 наиболее вероятных токенов, что уменьшает вероятность повторения. Вроде как переводы должны стать разнообразнее и уменьшится повторения.

yagera committed:
add: top-k sampling to translate for improved diversity

yagera committed:
feat: add new vocabs (50k tokens)

yagera committed:
feat: add new tokens (special and numbers)

yagera committed:
refactor: modular decoder layer

yagera committed:
fix: apply proper masking before softmax in multi-head attention

yagera committed:
feat: implement residual connection with pre-layer normalization

yagera committed:
fix: generate correct encoder and decoder masks with causal masking

yagera committed:
fix: refactor Transformer model layers and positional encoding

yagera committed:
fix: fix tokenization and vocabulary handling

yagera committed:
feat: improve dataset handling and translation efficiency

yagera committed:
feat: update training and model configuration

yageraaa commented:
> SwiGLU, RMSnorm, maybe QK-norm maybe RoPE

> после применения этих изменений можно сделать претрейне ТОЛЬКО декодера на английском языке. Претрейн работать будет по тому же принципе, что и основное обучение, но нету src_text. Декодер должен пробовать угадывать tgt_text без src. Метрику при претрейне считать не надо, модель просто должна получить первичные знания о языке

yagera committed:
feat(normalization): implement RMSNorm as an efficient alternative to LayerNorm

yagera committed:
feat(activation): add SwiGLU activation function to replace GELU in feed-forward layers

yagera committed:
feat(attention): implement QK-normalization for improved attention mechanism stability

yagera committed:
feat(encoder-block): implement enhanced encoder block with RMSNorm, SwiGLU and QK-norm

yagera committed:
feat(decoder-block): implement enhanced decoder block with RMSNorm, SwiGLU and QK-norm

yagera committed:
feat(encoder): integrate RMSNorm, SwiGLU and QK-norm into encoder architecture

yagera committed:
feat(decoder): integrate RMSNorm, SwiGLU and QK-norm into decoder architecture

yagera committed:
feat(transformer): implement enhanced transformer with RMSNorm, SwiGLU and QK-norm techniques

yagera committed:
feat(residual): implement residual connection with RMSNorm for improved gradient flow

yagera committed:
refactor(training): update training script to use enhanced transformer architecture

yageraaa commented:
наверное стоит разделить /models на две директории: transformer и transformer_v2

yagera committed:
feat(data): create dataset for English language model pretraining

yagera committed:
feat(model): implement decoder-only model for language modeling

yagera committed:
feat(training): implement decoder-only pretraining pipeline

yageraaa commented:
Я почитал ещё недавно про перплексию (метрику), которая показывает, насколько хорошо модель предсказывает последовательность слов или токенов. В самом обучении трансформера в ней толку нет, а вот например для претрейна декодера в будущем было бы неплохо или нет ?

yageraaa commented:
А также читал про другие метрики, которые часто используют для оценки перевода:

- TER (Translation Edit Rate) — оценивает минимальное количество операций редактирования, необходимых для превращения машинного перевода в эталонный, что отражает усилия по постредактированию.

- METEOR — учитывает точные совпадения, синонимы и однокоренные слова, а также баланс precision и recall, что позволяет лучше оценивать смысловую близость перевода.

- COMET — нейросетевая метрика, которая оценивает качество перевода с учётом беглости, адекватности и сохранения смысла, обучена на данных с человеческими оценками и даёт более глубокую оценку качества.

- chrF3 — основана на подсчёте совпадений символов и n-грамм, полезна для языков с морфологической сложностью.

- RIBES — учитывает порядок слов и ранговую корреляцию n-грамм, что важно для языков с разным синтаксисом, например, для русского и английского.

yagera committed:
feat(config_wmt): add parquet dataset and pretrained decoder support

yagera committed:
feat(data): implement new(wmt19) dataset loader for transformer training

yagera committed:
feat(data): implement new(wmt19) dataset loader for transformer training

yagera committed:
feat(transformer): add new transformer implementation to use with new dataset and new training pipeline

yageraaa commented:
датасет загружается нормально
по кол ву предложений всё сходится
```
Downloading data: 100%|██████████| 28/28 [02:27<00:00,  5.25s/files]
Generating train split: 100%|██████████| 37492126/37492126 [01:29<00:00, 420967.15 examples/s]
Generating validation split: 100%|██████████| 3000/3000 [00:00<00:00, 16398.41 examples/s]
```

yagera committed:
fix(load): add weights_only=True to safe decoder checkpoint loading

yagera committed:
feat(decoder_model): add cross-attention block to decoder layer

yageraaa commented:
============================= test session starts ==============================
collecting ... collected 2 items

test_compatibility.py::test_parameter_compatibility PASSED               [ 50%]============================================================
ТЕСТ 100% СОВПАДЕНИЯ ПАРАМЕТРОВ
============================================================
Параметры претрейн блока: 25
Параметры файн-тюнинг блока: 25
КЛЮЧИ ПАРАМЕТРОВ СОВПАДАЮТ 100%!
РАЗМЕРЫ ПАРАМЕТРОВ СОВПАДАЮТ 100%!
Общее количество параметров:
  Претрейн блок: 1,315,840
  Файн-тюнинг блок: 1,315,840
ОБЩЕЕ КОЛИЧЕСТВО ПАРАМЕТРОВ СОВПАДАЕТ 100%!

test_compatibility.py::test_weight_transfer PASSED                       [100%]
============================================================
ТЕСТ ПЕРЕНОСА ВЕСОВ
============================================================
Устройство: cuda

1. Инициализация модели для претрейна
Модель создана: 3,144,936 параметров

2. Мокнутое обучение первого этапа (5 шагов)
  Шаг 1/5, Loss: 7.1351
  Шаг 2/5, Loss: 7.1404
  Шаг 3/5, Loss: 7.1452
  Шаг 4/5, Loss: 7.1212
  Шаг 5/5, Loss: 7.1105
Веса изменились в процессе обучения

3. Сохранение модели
Модель сохранена: 12.08 MB

4. Инициализация новой модели

5. Загрузка весов
Веса загружены успешно
Загруженные веса совпадают с сохраненными

6. Мокнутое обучение второго этапа (5 шагов)
  Шаг 1/5, Loss: 6.9055
  Шаг 2/5, Loss: 6.9940
  Шаг 3/5, Loss: 7.0343
  Шаг 4/5, Loss: 7.0620
  Шаг 5/5, Loss: 7.0998
Веса изменились во втором этапе обучения

ТЕСТ ПЕРЕНОСА ПРОЙДЕН!

yagera committed:
fix: pass model argument to accelerator.get_state_dict

yagera committed:
feat(wandb): add support for resuming training in the same run

yagera committed:
feat(config): add model weights paths

yagera committed:
refactor(project-structure): reorganize files into directories
and add support for resuming training

yagera committed:
refactor(project-structure): reorganize files into directories

yagera committed:
feat(docker): add docker-compose configuration for postgres and minio services

yagera committed:
feat(backend): add FastAPI application with user and translation endpoints

yagera committed:
feat(backend): add SQLAlchemy database configuration and models

yagera committed:
feat(backend): add Pydantic models for API request and response validation

yagera committed:
feat(backend): add S3 client for file storage in MinIO

yagera committed:
feat(backend): add settings configuration with environment variable loading

yagera committed:
feat(backend): add utility functions for password hashing, JWT, and text translation

yagera committed:
feat(frontend): add Streamlit frontend for user authentication and translation

KORALLLL commented on api/backend/utils.py:
```diff
@@ -0,0 +1,64 @@
+    return encoded_jwt
+
+CONFIG_PATH = "/home/gera/PycharmProjects/russian-english-transformer/models/configs/config_wmt.yaml"
+cfg = OmegaConf.load(CONFIG_PATH)
```
Мне очень не нравится, что для конфига здесь задаётся полный путь. Можно  относительный, тк. этот конфиг в репо находитс

KORALLLL commented on api/backend/utils.py:
```diff
@@ -0,0 +1,64 @@
+            end_token_id=tokenizer.en_token_to_id['<end>'],
+            max_len=cfg.training.seq_len
+        )
+    return tokenizer.decode_ids(output[0].cpu().numpy(), tokenizer.en_id_to_token)
```
Вот это всё я бы хранился бы в папке с моделью. Как минимум, чтобы не дублировать. Либо, создай папку ml и там храни это

KORALLLL commented on models/core/residual_connection_rms.py:
объедини с предыдущим файлом. Выбор типа нормализации можно осуществлять через аргументы констрктора

KORALLLL commented on models/transformer/decoder_layer_v2.py:
Мы договорились, что старые блоки ты удалишь)

KORALLLL reviewed (changes requested):
чуть-чуть надо изменения сделать

yagera committed:
refactor(backend): use relative path for config_wmt.yaml to improve portability

yagera committed:
feat(ml): add translator module for ML translation logic in api/ml

yagera committed:
refactor(backend): remove ML logic from utils to separate api/ml/translator module

yagera committed:
fix(backend): update translate_text import to use new api.ml.translator module

yagera committed:
feat(docker): add Dockerfile for backend with ML support

yagera committed:
feat(docker): add Dockerfile for frontend

yagera committed:
feat(docker): add backend and frontend services to docker-compose

yagera committed:
chore(config): update .gitignore to exclude .env and temporary files

yagera committed:
chore(backend): update requirements.txt with dependencies for containerized backend

yagera committed:
chore(frontend): update requirements.txt with dependencies for containerized frontend

yagera committed:
fix(backend): add bcrypt and cryptography to passlib and python-jose dependencies

yagera committed:
feat(frontend): add streamlit-cookies-manager to requirements.txt

yagera committed:
chore(config): add unified requirements.txt with all project dependencies

yagera committed:
feat: add support for selectable normalization type (LayerNorm or RMSNorm)

yagera committed:
chore: remove file after merging into residual_connection.py

yagera committed:
chore: update imports in to include LayerNormalization and RMSNorm after file merge

yagera committed:
feat(translation): display original and translated text after translation

yagera committed:
refactor(backend/docker): copy only backend requirements instead of full source

yagera committed:
refactor(frontend/docker): streamline Dockerfile to copy only frontend requirements

yagera committed:
refactor(docker-compose): restructure services and volumes, update build contexts and env management

yagera committed:
chore(backend): update requirements.txt with additional dependencies

yagera committed:
refactor(backend): improve checkpoint loading and path handling in translator.py

yagera committed:
fix(config): require explicit AWS host/env values in settings

yagera committed:
chore(config): change base_dir from local dev path to /app/api for container compatibility

yagera committed:
feat: set default norm_type to rms in residual_connection.py

yagera committed:
feat: update transformer module by removing v2 and wmt files and delete old modules

yagera committed:
docs: add README with updated model weights instructions and project structure

yagera committed:
feat: add use_pretrained_decoder option in config.yaml and train.py for flexible decoder weights loading

yagera committed:
Merge remote-tracking branch 'origin/dev' into dev

yagera committed:
fix: fix training loop bug and correct config settings

yagera committed:
docs: add MIT License and update README with license section

KORALLLL commented on api/backend/Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM python:3.10-slim
+WORKDIR /app
+COPY api/backend/requirements.txt .
+RUN pip install --no-cache-dir -r requirements.txt
```
сделай через ```uv pip install ...``` так немного быстрее все собираться будет

KORALLLL commented on api/frontend/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM python:3.10-slim
+WORKDIR /app
+COPY api/frontend/requirements.txt ./requirements.txt
+RUN pip install --no-cache-dir -r requirements.txt
```
здесь и далее настоятельно советую перейти на uv для резолва зависимостей

KORALLLL reviewed (approved):
все хорошо,

yagera committed:
fix: change base_dir to avoid problems with paths

yagera committed:
fix: add entity to wandb initialization

yagera committed:
fix: change entity to wandb initialization

yagera committed:
fix: change back entity to wandb initialization

yagera committed:
feat: add the option to use MLFlow instead of Wandb

yagera committed:
fix(bleu): add text normalization before comparison

yagera committed:
fix: delete an empty line

yagera committed:
fix: replace the mention of Wandb with MLFlow

yagera committed:
feat: increase batch size

yagera committed:
fix: reduce batch size

yagera committed:
feat: increase batch size

yagera committed:
feat: add possibility to train model on two cards

yagera committed:
fix: return changes

yagera committed:
feat: add cash deleting after train stop

yagera committed:
feat: increase batch size

yagera committed:
feat: add possibility to use 2 cards

yagera committed:
feat: again add possibility to use 2 cards

yagera committed:
feat: revert changes

yagera committed:
feat: increase batch size to 256

yagera committed:
feat: reduce batch size to 128

yagera committed:
fix: log translations

yagera committed:
fix: reduce validation interval

yagera committed:
fix: increase validation interval

yagera committed:
fix: increase validation interval

yagera committed:
fix: launch via accelerate

yagera committed:
fix: change seq_len

yagera committed:
refactor: change parameters in config

yagera committed:
fix: make some changes in masking, weight initialization, attention and max_len in pos coding

yagera committed:
fix: revert config and implement num workers in training

yagera committed:
fix: revert config and implement num workers in training

yagera committed:
refactor: rolled back the changes pending approval

yagera committed:
refactor: feat: restore original model config and add num_workers=48

yagera committed:
fix: add gradient clipping, fix BLEU metric calculation, correct loss and decoding tokens, optimize dataloader workers for A100 GPUs

yagera committed:
refactor: reduce val interval steps

yagera committed:
refactor: increase val interval steps

yagera committed:
feat: make improvements that should stabilize learning and also added a new metric

yagera committed:
refactor: reduce batch size

yagera committed:
fix: make some changes

yagera committed:
fix: make changes regarding the metrics

yagera committed:
refactor: delete METEOR metric:(

yagera committed:
refactor: increase val steps

yagera committed:
refactor: reduce val steps

yagera committed:
refactor: increase val steps back

yagera committed:
fix: use weights_only=False to allow OmegaConf objects in the checkpoint

yagera committed:
fix: make some changes in load_state_dict method

yagera committed:
refactor(config): reduce model parameters

yagera committed:
feat: implement bf16 training and robust error handling for NCCL timeout issues

yagera committed:
feat: increase batch size

yagera committed:
fix: update deprecated PyTorch AMP API to modern syntax

yagera committed:
fix: resolve multi-GPU training stability issues

yagera committed:
fix: resolve NCCL timeout issues in multi-GPU training

yagera committed:
refactor(training): prevent NCCL timeouts by adjusting Accelerator and DataLoader usage

yagera committed:
refactor(training): prevent NCCL timeouts by adjusting Accelerator and DataLoader usage

yagera committed:
refactor(config): increase learning rate to 3e-4

yagera committed:
refactor(config): increase model size to 32M

yagera committed:
refactor: reduce batch size
